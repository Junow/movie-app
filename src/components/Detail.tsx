import React from 'react'
import styled from 'styled-components'
import {fetcher} from '../common/request'
import {useParams} from 'react-router-dom'
import * as T from '../types/api'
import {getYear, convertMinToHour} from '../common/utils'
import useSwr from 'swr'
import LazyImage from './common/LazyImage'

const DetailContainer = styled.section<{backdrop: string}>`
  display: flex;
  flex-direction: column;

  .poster {
    width: 20rem;
    height: auto;
    border-radius: 1rem;
  }

  .detail-info {
    margin: 3rem;
    > h1 {
      font-size: 2rem;
    }
    > div {
      span {
        &:not(:first-child)::before{
          content: ' • ';
        }
      }
    }
    .tagline {
      color: #BFC0C6;
      font-weight: 500;
    }
  }

  .detail-header{
    display: flex;
    width: 100%;
    max-width: 1300px;

    padding: 1rem;
  }

  .background-overlay{
    display: flex;
    width: 100%;
    justify-content: center;
    align-items: center;

    background-image: linear-gradient(to right, rgba(12.94%, 14.90%, 22.75%, 1.00) 150px, rgba(20.39%, 22.35%, 29.02%, 0.84) 100%);
  }
  .background{
    ${(props) => `background-image: url(${props.backdrop})`};
    background-repeat: no-repeat;
    background-position: top center;
    background-size: cover;
  }

  .cast-container {
    max-width: 1300px;
    overflow-x: scroll;
    margin: 0 auto;
  }
  
  .cast-list {
    display: flex;
    overflow-x: scroll;
  }

  .cast {
    &:not(:first-child) {
      margin-left: 1rem;
    }

    .cast-name {
      font-weight: 600;
    }
    .cast-character {
      font-weight: 300;
    }

  }
`

const getDetailUrl = (id: string) => `https://api.themoviedb.org/3/movie/${id}?language=ko-KR&append_to_response=credits`

export default function Detail(){
  const {id} = useParams<{id: string}>()
  const {data, error} = useSwr<T.Detail>(getDetailUrl(id), fetcher)

  if(error) {
    return <div>...error occured</div>
  }
  if(!data){
    return <div>...loading</div>
  }
  
  return (
    <DetailContainer backdrop={`https://image.tmdb.org/t/p/w1920_and_h800_multi_faces${data.backdrop_path}`}>
      <div className="background">
        <div className="background-overlay">
          <section className="detail-header">
            <img className="poster" src={`https://image.tmdb.org/t/p/w500${data.poster_path}`}/>
            <div className='detail-info'>
              <h1>{data.title} ({getYear(data.release_date)})</h1>
              <div>
                <span>{data.release_date}</span>
                <span>{data.genres.map(genre => genre.name).join(', ')}</span>
                {
                  data.runtime && <span>{convertMinToHour(data.runtime)}</span>
                }
              </div>
              <h3>평가</h3>
              <span>{data.vote_average}({data.vote_count})</span>
              <h4 className="tagline"><i>{data.tagline}</i></h4>
              <h3>개요</h3>
              <p>
                {data.overview}
              </p>
            </div>
          </section>
        </div>
      </div>
      <section className='cast-container'> 
        <h1>주요 출연진</h1>
        <div className='cast-list'> 
          {
            data.credits.cast.map(cast => {
              return (
                <div className='cast'>
                  <LazyImage src={`https://image.tmdb.org/t/p/w276_and_h350_face${cast.profile_path}`}/>
                  <div>
                    <div className='cast-name'>{cast.name}</div>
                    <div className='cast-character'>{cast.character}</div>
                  </div>
                </div>
              )
            })
          }
        </div>
      </section>
    </DetailContainer>
  )
}
