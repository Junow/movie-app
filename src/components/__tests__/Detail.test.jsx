import React, { ReactChild, ReactNode } from 'react'
import {render} from '@testing-library/react'
import Detail from '../Detail'
import {createMemoryHistory, MemoryHistory, History} from 'history'
import { Router, Route, MemoryRouter } from "react-router-dom";
describe('<Detail/>', () => {

  const setup = () => {
    return render(<Detail/>)
  }

  function renderWithRouter(
    ui,
    {
      route = '/',
      history = createMemoryHistory({ initialEntries: [route] }),
    } = {}
  ) {
    const Wrapper = ({ children }) => (
      <Router history={history}>{children}</Router>
    )
    return {
      ...render(ui, { wrapper: Wrapper }),
      // adding `history` to the returned utilities to allow us
      // to reference it in our tests (just try to avoid using
      // this to test implementation details).
      history,
    }
  }


  it('should render properly', () => {
    const route = '/detail/123'
    const wrapper = renderWithRouter(<Detail />, { route }) 
    expect(wrapper).toMatchSnapshot()
  })
  

})
