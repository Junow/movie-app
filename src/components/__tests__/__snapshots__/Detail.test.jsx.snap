// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Detail/> should render properly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        ...loading
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      ...loading
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "history": Object {
    "action": "POP",
    "block": [Function],
    "canGo": [Function],
    "createHref": [Function],
    "entries": Array [
      Object {
        "hash": "",
        "key": "x76vbl",
        "pathname": "/detail/123",
        "search": "",
        "state": undefined,
      },
    ],
    "go": [Function],
    "goBack": [Function],
    "goForward": [Function],
    "index": 0,
    "length": 1,
    "listen": [Function],
    "location": Object {
      "hash": "",
      "key": "x76vbl",
      "pathname": "/detail/123",
      "search": "",
      "state": undefined,
    },
    "push": [Function],
    "replace": [Function],
  },
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
